File: src/test/java/legacy/MoneyTest.java
package legacy;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

import java.math.BigDecimal;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;

public class MoneyTest {

    @Test
    void shouldThrowExceptionForNullValue() {
        // when
        Executable whenBlock = () -> Money.of(null, Currency.PLN);

        // then
        assertThrows(IllegalStateException.class, whenBlock);
    }

    @Test
    void shouldThrowExceptionForNullCurrency() {
        // when
        Executable whenBlock = () -> Money.of(BigDecimal.TEN, null);

        // then
        assertThrows(IllegalStateException.class, whenBlock);
    }

    @Test
    void shouldAddTwoValues() {
        // given
        var oneEur = Money.of(BigDecimal.valueOf(1), Currency.EUR);
        var tenEur = Money.of(BigDecimal.valueOf(10), Currency.EUR);

        // when
        var result = oneEur.plus(tenEur);

        // then
        assertThat(result.getValue()).isEqualTo(BigDecimal.valueOf(11));
        assertThat(result.getCurrency()).isEqualTo(Currency.EUR);
    }

    @Test
    void shouldSubtractTwoValues() {
        // given
        var oneEur = Money.of(BigDecimal.valueOf(1), Currency.EUR);
        var tenEur = Money.of(BigDecimal.valueOf(10), Currency.EUR);

        // when
        var result = tenEur.minus(oneEur);

        // then
        assertThat(result.getValue()).isEqualTo(BigDecimal.valueOf(9));
        assertThat(result.getCurrency()).isEqualTo(Currency.EUR);
    }

    @Test
    void plusOperationIsCommutative() {
        // given
        var oneEur = Money.of(BigDecimal.valueOf(1), Currency.EUR);
        var tenEur = Money.of(BigDecimal.valueOf(10), Currency.EUR);

        // when
        var firstResult = oneEur.plus(tenEur);
        var secondResult = tenEur.plus(oneEur);

        // then
        assertThat(firstResult).isEqualTo(secondResult);
    }

    @Test
    void shouldThrowForPlusOperationWhenCurrenciesDiffer() {
        // given
        var oneEur = Money.of(BigDecimal.valueOf(1), Currency.EUR);
        var onePln = Money.of(BigDecimal.valueOf(1), Currency.PLN);

        // when
        Executable whenBlock = () -> oneEur.plus(onePln);

        // then
        assertThrows(IllegalStateException.class, whenBlock);
    }

    @Test
    void shouldMultiplyByPercent() {
        // given
        var tenEur = Money.of(BigDecimal.valueOf(10), Currency.PLN);
        var threePercent = Percent.of(BigDecimal.valueOf(3.5));

        // when
        var result = tenEur.times(threePercent);

        // then
        assertThat(result.getValue()).isEqualTo(BigDecimal.valueOf(0.35));
        assertThat(result.getCurrency()).isEqualTo(Currency.PLN);
    }
}

File: src/test/kotlin/refactored/MoneyTest.kt
package refactored

import assertk.assertThat
import assertk.assertions.isEqualTo
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.math.BigDecimal

class MoneyTest {

    @Test
    fun `should add two values`() {
        // given
        val oneEur = Money(BigDecimal.valueOf(1), Currency.EUR)
        val tenEur = Money(BigDecimal.valueOf(10), Currency.EUR)

        // when
        val result = oneEur.plus(tenEur)

        // then
        assertThat(result.value).isEqualTo(BigDecimal.valueOf(11))
        assertThat(result.currency).isEqualTo(Currency.EUR)
    }

    @Test
    fun `should subtract two values`() {
        // given
        val oneEur = Money(BigDecimal.valueOf(1), Currency.EUR)
        val tenEur = Money(BigDecimal.valueOf(10), Currency.EUR)

        // when
        val result = tenEur.minus(oneEur)

        // then

        assertThat(result.value).isEqualTo(BigDecimal.valueOf(9))
        assertThat(result.currency).isEqualTo(Currency.EUR)
    }

    @Test
    fun plusOperationIsCommutative() {
        // given
        val oneEur = Money(BigDecimal.valueOf(1), Currency.EUR)
        val tenEur = Money(BigDecimal.valueOf(10), Currency.EUR)

        // when
        val firstResult = oneEur.plus(tenEur)
        val secondResult = tenEur.plus(oneEur)

        // then
        assertThat(firstResult).isEqualTo(secondResult)
    }

    @Test
    fun shouldThrowForPlusOperationWhenCurrenciesDiffer() {
        // given
        val oneEur = Money(BigDecimal.valueOf(1), Currency.EUR)
        val onePln = Money(BigDecimal.valueOf(1), Currency.PLN)

        // when
        val whenBlock = {
            oneEur.plus(
                onePln
            )
        }

        // then
        assertThrows<IllegalStateException>{ whenBlock() }
    }

    @Test
    fun shouldMultiplyByPercent() {
        // given
        val tenEur = Money(BigDecimal.valueOf(10), Currency.PLN)
        val threePercent = Percent(BigDecimal.valueOf(3.5))

        // when
        val result = tenEur.times(threePercent)

        // then
        assertThat(result.value).isEqualTo(BigDecimal.valueOf(0.35))
        assertThat(result.currency).isEqualTo(Currency.PLN)
    }
}